### PR TITLE
Make `likely` and `unlikely` const, gated by feature `const_unlikely`

### PR DESCRIPTION
**Pull request by nbdd0121**
This PR also contains a fix to allow `#[allow_internal_unstable]` to work properly with `#[rustc_const_unstable]`.

cc @RalfJung @nagisa 

r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/interpret/intrinsics.rs [hunk 1]
```diff
@@ -399,6 +399,11 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 );
                 self.copy_op(self.operand_index(args[0], index)?, dest)?;
             }
+            sym::likely | sym::unlikely => {
+                // These just return their argument
+                let a = self.read_immediate(args[0])?;
+                self.write_immediate(*a, dest)?;
```
Comment (oli-obk) [inline src/librustc_mir/interpret/intrinsics.rs:405]: I believe

```suggestion
                self.copy_op(args[0], dest)?;
```

will do the same thing, but with more checks.

**Comment by oli-obk** [comment]
lgtm. I'll add it to the const eval skill tree.

@rust-lang/lang This PR does two things:

1. A trivial change making the `likely` and `unlikely` intrinsics unstably const fn. This means only unstable const fns can invoke these intrinsics. Since there's no way for a stable const fn to invoke an unstable const fn, that would be an invisible change. Additionally, even if we stabilized the constness of these intrinsics, they are only performance relevant, so we can destabilize them and take performance hits if we ever wanted to (I don't know why, these intrinsics are completely benign from a const eval perspective, they are nops).
2. Allow *stable* const fn in the libstd to invoke *unstable* const fn iff the stable const fn has an `allow_internal_unstable(the_const_feature_gate_of_the_unstable_const_fn)` attribute. See https://github.com/rust-lang/rustc-dev-guide/blob/master/src/stability.md#allow_internal_unstable for documentation on the use of `allow_internal_unstable` on `const fn`.

The motivation for this change is to allow the [`checked_neg` function to compile to assembly preferring the happy path](https://github.com/rust-lang/rust/issues/73731).

https://github.com/rust-lang/rust/blob/394e1b40d264aa6928811919c1124fa248e7d802/src/libcore/num/mod.rs#L1867-L1877

`checked_neg` is the only const-stable `checked_*` function, because it used to be implemented by math trickery instead of an `if/else` expression. We moved to an `if/else` expression allowing `const_if_match` via `allow_internal_unstable`, because we could always move back to the math trickery if there would be any problems. In order to keep it stable and using `if/else`, we need to allow the use of the `unlikely` intrinsic within this const-stable function. This is the first time we've permitted calling a const-unstable intrinsic from a const-stabale function, thus the request for your sign-off on the new capabilities of `allow_internal_unstable`.

Alternatively we can insta-stabilize the const-stability of the `likely` and `unlikely` intrinsics, but there is a [general desire for such `allow_internal_unstable` capabilities](https://github.com/rust-lang/rust/issues/69399).

**Comment by RalfJung** [comment]
Please remember to [remove the intrinsic implementation from Miri](https://github.com/rust-lang/miri/blob/76d99cfa68d96a591e3e132cf73ef750f09d1a8a/src/shims/intrinsics.rs#L510) once this one lands.

File: src/librustc_mir/interpret/intrinsics.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/intrinsics.rs:405]: I think the amount of checking is the same... except the `immediate` code also requires that the value be an immediate, whereas `copy_op` also works with non-immediate values.

I do not have a strong preference.

**Comment by nagisa** [comment]
> The motivation for this change is to allow the checked_neg function to compile to assembly preferring the happy path.

Note, the linked issue is about `checked_div`, not `neg`.

**Comment by nbdd0121** [comment]
@oli-obk Note that this PR does not fix #69399, because
https://github.com/rust-lang/rust/blob/8b43012453de32d4e429c923171131f938d3ead8/src/librustc_mir/transform/qualify_min_const_fn.rs#L338-L343
which is copied from 
https://github.com/rust-lang/rust/blob/8b43012453de32d4e429c923171131f938d3ead8/src/librustc_mir/transform/qualify_min_const_fn.rs#L318-L323
initially introduced in #68114.

This is different from the behaviour of `#[allow_internal_unstable]` behaviour when applied to macros today, which does not require the feature to be enabled on crate level. Is there a reason for the restriction @ecstatic-morse ?

**Comment by ecstatic-morse** [comment]
What's the question exactly? I'm not sure what you want to change.

**Comment by nbdd0121** [comment]
@ecstatic-morse `#[allow_internal_unstable]` requires the feature gate to be enabled on the crate when applied to functions. I just want to know the reason behind this restriction because you introduced it, and ideally have the restriction removed because this the restriction does not apply when `#[allow_internal_unstable]` is used on `macro_rules!`.

**Comment by ecstatic-morse** [comment]
I believe this was just the way it worked before, although this was broken for a while. Whats the problem with adding your library feature gate to `libcore`?

**Comment by nbdd0121** [comment]
> I believe this was just the way it worked before, although this was broken for a while. Whats the problem with adding your library feature gate to `libcore`?

It's not an issue for me, but it is blocking #69399. And it is just weird that the same attribute has different restrictions when applied to function/macro respectively.

**Comment by ecstatic-morse** [comment]
I don't think the current behavior is weird. Writing a macro definition is not the same as writing a function. I also don't care very much.

**Comment by nikomatsakis** [comment]
I'm nominating for the @rust-lang/lang meeting today but I don't see a problem with having these functions available in const fn, presuming that we have them. They clearly have no semantic effect. This is only to make them available on an unstable basis so I think no FCP ought to be required.

**Comment by scottmcm** [comment]
I agree that unstable things generally don't need a lang FCP, especially for intrinsics that are by nature internal and more about compiler/libs-impl concerns than about the conceptual language itself.

**Comment by nikomatsakis** [comment]
Discussed this PR and gave approval in @rust-lang/lang meeting today.

**Pull request merged**